### DISCUSSION TITLE
unble to link curl object files

### DISCUSSION THREAD

**Discussion by Dhruvan117** [curl/curl#12354] [Q&A] [answered]
### I did this

```
D:\play\curl-master>mingw32-make mingw32-zlib
mingw32-make -C lib -f Makefile.mk CFG=mingw32-zlib
mingw32-make[1]: Entering directory 'D:/play/curl-master/lib'
mingw32-make[1]: Nothing to be done for 'all'.
mingw32-make[1]: Leaving directory 'D:/play/curl-master/lib'
mingw32-make -C src -f Makefile.mk CFG=mingw32-zlib
INFO: Could not find files for the given pattern(s).
mingw32-make[1]: Entering directory 'D:/play/curl-master/src'
gcc   -L"../../zlib/lib" -L../lib  -o curl.exe mingw32/slist_wc.o mingw32/tool_binmode.o mingw32/tool_bname.o mingw32/tool_cb_dbg.o mingw32/tool_cb_hdr.o mingw32/tool_cb_prg.o mingw32/tool_cb_rea.o mingw32/tool_cb_see.o mingw32/tool_cb_wrt.o mingw32/tool_cfgable.o mingw32/tool_dirhie.o mingw32/tool_doswin.o mingw32/tool_easysrc.o mingw32/tool_filetime.o mingw32/tool_findfile.o mingw32/tool_formparse.o mingw32/tool_getparam.o mingw32/tool_getpass.o mingw32/tool_help.o mingw32/tool_helpers.o mingw32/tool_hugehelp.o mingw32/tool_ipfs.o mingw32/tool_libinfo.o mingw32/tool_listhelp.o mingw32/tool_main.o mingw32/tool_msgs.o mingw32/tool_operate.o mingw32/tool_operhlp.o mingw32/tool_paramhlp.o mingw32/tool_parsecfg.o mingw32/tool_progress.o mingw32/tool_setopt.o mingw32/tool_sleep.o mingw32/tool_stderr.o mingw32/tool_strdup.o mingw32/tool_urlglob.o mingw32/tool_util.o mingw32/tool_vms.o mingw32/tool_writeout.o mingw32/tool_writeout_json.o mingw32/tool_xattr.o mingw32/var.o mingw32/base64.o mingw32/curl_multibyte.o mingw32/dynbuf.o mingw32/nonblock.o mingw32/strtoofft.o mingw32/timediff.o mingw32/version_win32.o mingw32/warnless.o -lcurl   -lz
mingw32/slist_wc.o:slist_wc.c:(.text+0xc9): undefined reference to `_imp__curl_slist_append'
mingw32/slist_wc.o:slist_wc.c:(.text+0x101): undefined reference to `_imp__curl_slist_free_all'
mingw32/slist_wc.o:slist_wc.c:(.text+0x159): undefined reference to `_imp__curl_slist_free_all'
mingw32/tool_cb_dbg.o:tool_cb_dbg.c:(.text+0x140): undefined reference to `_imp__curl_msnprintf'
mingw32/tool_cb_dbg.o:tool_cb_dbg.c:(.text+0x1a6): undefined reference to `_imp__curl_mfprintf'
mingw32/tool_cb_dbg.o:tool_cb_dbg.c:(.text+0x231): undefined reference to `_imp__curl_msnprintf'
mingw32/tool_cb_dbg.o:tool_cb_dbg.c:(.text+0x275): undefined reference to `_imp__curl_easy_getinfo'
mingw32/tool_cb_dbg.o:tool_cb_dbg.c:(.text+0x2b6): undefined reference to `_imp__curl_easy_getinfo'
[edit: same for all exported curl symbols]
mingw32/var.o:var.c:(.text+0x504): undefined reference to `_imp__curl_free'
mingw32/var.o:var.c:(.text+0xcca): undefined reference to `_imp__curl_getenv'
mingw32/var.o:var.c:(.text+0xe82): undefined reference to `_imp__curl_free'
mingw32/dynbuf.o:dynbuf.c:(.text+0x34f): undefined reference to `_imp__curl_mvaprintf'
mingw32/nonblock.o:nonblock.c:(.text+0xe5): undefined reference to `ioctlsocket@12'
collect2.exe: error: ld returned 1 exit status
Makefile.mk:109: recipe for target 'curl.exe' failed
mingw32-make[1]: *** [curl.exe] Error 1
mingw32-make[1]: Leaving directory 'D:/play/curl-master/src'
Makefile:51: recipe for target 'mingw32-zlib' failed
mingw32-make: *** [mingw32-zlib] Error 2
```

### curl/libcurl version

8.4.0

### operating system

windows 11

**Comment by vszakats** [comment]
A semi-blind guess, but it looks as if Windows-autodetection didn't work in `lib/Makefile.mk` (it's based on `gcc -dumpmachine` returning a value with `-w` in it), then missed to set `-DCURL_STATICLIB` and add the necessary Windows systems libs.

You can work this around by passing `WIN32=1` to make. Having said that, `Makefile.mk` will be removed (#12224) after the upcoming release, so it may even be better to try building with CMake.

What does the command `gcc -dumpmachine` return on your system?

UPDATE: One possible explanation for what's happening is that this `gcc` is _not_ a mingw gcc, but an MSYS/Cygwin one, which isn't Windows and thus won't work with this build method (it's only for Windows, MS-DOS and Amiga). In that case the fix is passing `CC=x86_64-w64-mingw32-gcc` (or `CC=i686-w64-mingw32-gcc`) to make.

**Comment by Dhruvan117** [comment]
all:
	./configure
	make

ssl:
	./configure --with-openssl
	make

mingw32:
	$(MAKE) -C lib -f Makefile.mk
	$(MAKE) -C src -f Makefile.mk

mingw32-clean:
	$(MAKE) -C lib -f Makefile.mk clean
	$(MAKE) -C src -f Makefile.mk clean
	$(MAKE) -C docs/examples -f Makefile.mk clean

mingw32-vclean mingw32-distclean:
	$(MAKE) -C lib -f Makefile.mk vclean
	$(MAKE) -C src -f Makefile.mk vclean
	$(MAKE) -C docs/examples -f Makefile.mk vclean

mingw32-examples%:
	$(MAKE) -C docs/examples -f Makefile.mk CFG=$@

mingw32%:
	$(MAKE) -C lib -f Makefile.mk CFG=$@
	$(MAKE) -C src -f Makefile.mk CFG=$@

vc:
	cd winbuild
	nmake /f Makefile.vc MACHINE=x86

vc-x64:
	cd winbuild
	nmake /f Makefile.vc MACHINE=x64

djgpp%:
	$(MAKE) -C lib -f Makefile.mk CFG=$@ CROSSPREFIX=i586-pc-msdosdjgpp-
	$(MAKE) -C src -f Makefile.mk CFG=$@ CROSSPREFIX=i586-pc-msdosdjgpp-

cygwin:
	./configure
	make

cygwin-ssl:
	./configure --with-openssl
	make

amiga%:
	$(MAKE) -C lib -f Makefile.mk CFG=$@ CROSSPREFIX=m68k-amigaos-
	$(MAKE) -C src -f Makefile.mk CFG=$@ CROSSPREFIX=m68k-amigaos-

unix: all

unix-ssl: ssl

linux: all

linux-ssl: ssl

ca-bundle: scripts/mk-ca-bundle.pl
	@echo "generate a fresh ca-bundle.crt"
	@perl $< -b -l -u lib/ca-bundle.crt

ca-firefox: lib/firefox-db2pem.sh
	@echo "generate a fresh ca-bundle.crt"
	./lib/firefox-db2pem.sh lib/ca-bundle.crt
	
This is my main make file, it was automatically generated when i run buidconfirm windows batch file.


PROOT := ..

### Common

include $(PROOT)/lib/Makefile.mk

### Local

RCFLAGS  += -DCURL_EMBED_MANIFEST
CPPFLAGS += -I$(PROOT)/lib
LDFLAGS  += -L$(PROOT)/lib
LIBS     := -lcurl $(LIBS)

ifdef WIN32
  ifneq ($(findstring -dyn,$(CFG)),)
    DYN := 1
  endif
endif

ifdef DYN
  curl_DEPENDENCIES := $(PROOT)/lib/libcurl$(CURL_DLL_SUFFIX).dll
  curl_DEPENDENCIES += $(PROOT)/lib/libcurl.dll.a
else
  curl_DEPENDENCIES := $(PROOT)/lib/libcurl.a
  ifdef WIN32
    CPPFLAGS += -DCURL_STATICLIB
    LDFLAGS += -static
  endif
endif

### Sources and targets

# Provides CURL_CFILES, CURLX_CFILES, CURL_RCFILES
include Makefile.inc

TARGETS := curl$(BIN_EXT)

CURL_CFILES += $(notdir $(CURLX_CFILES))

curl_OBJECTS := $(patsubst %.c,$(OBJ_DIR)/%.o,$(strip $(CURL_CFILES)))
ifdef WIN32
curl_OBJECTS += $(patsubst %.rc,$(OBJ_DIR)/%.res,$(strip $(CURL_RCFILES)))
endif
ifdef MAP
CURL_MAP := curl.map
CURL_LDFLAGS_BIN += -Wl,-Map,$(CURL_MAP)
TOVCLEAN := $(CURL_MAP)
endif
vpath %.c $(PROOT)/lib

TOCLEAN := $(curl_OBJECTS)

### Rules

ifneq ($(wildcard tool_hugehelp.c.cvs),)
PERL  ?= perl
NROFF ?= groff

TOCLEAN += tool_hugehelp.c

ifneq ($(shell $(call WHICH, $(NROFF))),)
$(PROOT)/docs/curl.1: $(wildcard $(PROOT)/docs/cmdline-opts/*.d)
	cd $(PROOT)/docs/cmdline-opts && \
	$(PERL) gen.pl mainpage $(notdir $^) > ../curl.1

# Necessary for the generated tools_hugehelp.c
CPPFLAGS += -DUSE_MANUAL

ifdef ZLIB
_MKHELPOPT += -c
endif
tool_hugehelp.c: $(PROOT)/docs/curl.1 mkhelp.pl
	$(NROFF) -man -Tascii $(MANOPT) $< | \
	$(PERL) mkhelp.pl $(_MKHELPOPT) $< > $@
else
tool_hugehelp.c:
	@echo Creating $@
	@$(call COPY, $@.cvs, $@)
endif
endif

$(TARGETS): $(curl_OBJECTS) $(curl_DEPENDENCIES)
	$(CC) $(LDFLAGS) $(CURL_LDFLAGS_BIN) -o $@ $(curl_OBJECTS) $(LIBS)

all: $(OBJ_DIR) $(TARGETS)

this is my source make file.
i am newbie in make files, Let me know if you want more details or more make file information.
Thanks in advance.

**Comment by vszakats** [comment]
Please read my previous reply for the possible solutions. This looks like wrong configuration on your end, where this mingw makefile is run with a non-mingw (native MSYS2) compiler. Try passing the correct `CC` value or switch to CMake.

**Comment by Dhruvan117** [comment]
D:\play\curl-master>gcc -dumpmachine
mingw32

![image](https://github.com/curl/curl/assets/99899247/9d9ad14a-322a-44cb-83b4-97a801285267)

tried hardcoded win32 macro enable by 1, still many files unavailable at linking time.
gcc -dumpmachine showing mingw32.

**Comment by vszakats** [comment]
It's an odd output for `-dumpmachine`; it is supposed to be a triplet. Where did you get this mingw-w64 installation from?

**Comment by Dhruvan117** [comment]
dowloaded mingw setup from this site : https://sourceforge.net/projects/mingw/

![image](https://github.com/curl/curl/assets/99899247/ceacc872-9a49-4953-8230-1ef461af7531)

**Comment by vszakats** [comment, marked as answer]
This is "old" mingw. I did not test these Makefiles with it and curl dropped support for it recently.

Try mingw-w64 from one of these sources: https://www.mingw-w64.org/. 

E.g. MSYS2, MinGW-W64-builds, or llvm-mingw from: https://github.com/mstorsjo/llvm-mingw/releases. I strongly suggest using CMake because as I said these Makefiles are deleted (for Windows) in 2-3 weeks.

**Comment by Dhruvan117** [comment]
In cmake i am getting this error, i will try new mingw compiler tool chain and will provide update here.
![image](https://github.com/curl/curl/assets/99899247/df92134f-3bf4-4918-9300-b5978b077802)

**Comment by vszakats** [comment]
This still looks like a very old mingw. I cannot see any curl issue here, so moving this to discussions.

**Comment by Dhruvan117** [comment]
https://github.com/mstorsjo/llvm-mingw/releases/download/20231114/llvm-mingw-20231114-msvcrt-x86_64.zip

This link worked for me (although i have to add mingw-make additionally and missing copy libcurl.dll file to the curl.exe folder).
I am able to build curl.exe without ssl.
Right now trying to make it with SSL support but getting an error in sslconfiguration. h header file missing.

but i think i will find a workaround of it and will be able to build a curl.exe executable soon with SSL support.

Thanks @vszakats for your support.
Have a great day.
I think you guys can place this mingw tool chain compiler link in window installation part of curl documantion, it would be helpfull to other people if they are getting same error as mine.

**Comment by vszakats** [comment]
llvm-mingw is a very nice binary package indeed; probably the best one at the moment. Nevertheless curl builds fine with any other mingw-w64 distro.

For SSL, the easiest is to start with Schannel, it's `-DCURL_USE_SCHANNEL=ON` with CMake.
